README.md
# Geoflram

[![Build Status](https://travis-ci.org/Colisweb/geoflram.svg?branch=master)](https://travis-ci.org/Colisweb/geoflram)
[![codecov](https://codecov.io/gh/Colisweb/geoflram/branch/master/graph/badge.svg)](https://codecov.io/gh/Colisweb/geoflram)

Installation
------------

Add the following to your `build.sbt` file:

```scala
resolvers += Resolver.bintrayRepo("colisweb", "maven")

libraryDependencies += "com.colisweb" %% "geoflram" % "1.1.2"
```

Use
---

```scala
import com.colisweb.geoflram.Geoflram
import com.colisweb.geoflram.Geoflram.Commune

val found: Option[Commune] = Geoflram.findBy(43.45603285, 0.5390568833333329)
```

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

More infos about this file : http://keepachangelog.com/

## [Unreleased] - no_due_date

- **Add OpenJDK11 in the Travis matrix**

## [v1.1.2] - 2018.12.14

- **Dependency updates**
  - Scala:  2.12.7 -> 2.12.8
  - sbt:    1.2.6 -> 1.2.7

## [v1.1.1] - 2018.11.22

- **Fix generated `.pom` file**

## [v1.1.0] - 2018.11.21

- **Add a `jruby` module which contains code optimized for use in JRuby**

## [v1.0.0] - 2018.11.21

- **Publish from `master` branch**

## [v0.1.4] - 2018.11.21

- **Fix publishing to Bintray**

## [v0.1.2] - 2018.11.21

- **Rename packages from `com.guizmaii.geofla` to `com.colisweb.geoflram`**
- **Rename object `Geogla` to `Geoflram`**

## [v0.1.1] - 2018.11.21

- **Precise the SRID used in loaded data (WGS84) in the `GeometryFactory`**

## [v0.1.0] - 2018.11.21

- **Init project**

core/src/test/scala/GeoflaSpec.scala
import com.guizmaii.geofla.Geofla
import org.scalatest.{FlatSpec, Matchers}

class GeoflaSpec extends FlatSpec with Matchers {

  "true" should "be true" in {
    true shouldBe true
  }

  "Geofla#findBy" should "find the correct commune" in {
    final case class GpsPoint(latitude: Double, longitude: Double)

    val `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France` =
      GpsPoint(latitude = 49.4296517, longitude = 1.2554941)

    val `Rue des Huttes, tertiary, Coingt, France` = GpsPoint(latitude = 49.7847695, longitude = 4.0952504)

    val `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France` =
      GpsPoint(latitude = 43.45603285, longitude = 0.5390568833333329)

    val london = GpsPoint(latitude = 51.5073219, longitude = -0.1276474)

    Geofla
      .findBy(
        `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France`.latitude,
        `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France`.longitude,
      )
      .map(_.nomCommune) shouldBe Some("BOIS-D'ENNEBOURG")

    Geofla
      .findBy(
        `Rue des Huttes, tertiary, Coingt, France`.latitude,
        `Rue des Huttes, tertiary, Coingt, France`.longitude,
      )
      .map(_.nomCommune) shouldBe Some("COINGT")

    Geofla
      .findBy(
        `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France`.latitude,
        `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France`.longitude,
      )
      .map(_.nomCommune) shouldBe Some("LOURTIES-MONBRUN")

    Geofla
      .findBy(
        london.latitude,
        london.longitude,
      )
      .map(_.nomCommune) shouldBe None
  }

  "Geofla#strTreefindBy" should "find the correct commune" in {
    final case class GpsPoint(latitude: Double, longitude: Double)

    val `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France` =
      GpsPoint(latitude = 49.4296517, longitude = 1.2554941)

    val `Rue des Huttes, tertiary, Coingt, France` = GpsPoint(latitude = 49.7847695, longitude = 4.0952504)

    val `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France` =
      GpsPoint(latitude = 43.45603285, longitude = 0.5390568833333329)

    val london = GpsPoint(latitude = 51.5073219, longitude = -0.1276474)

    Geofla
      .strTreefindBy(
        `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France`.latitude,
        `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France`.longitude,
      )
      .map(_.nomCommune) shouldBe Some("BOIS-D'ENNEBOURG")

    Geofla
      .strTreefindBy(
        `Rue des Huttes, tertiary, Coingt, France`.latitude,
        `Rue des Huttes, tertiary, Coingt, France`.longitude,
      )
      .map(_.nomCommune) shouldBe Some("COINGT")

    Geofla
      .strTreefindBy(
        `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France`.latitude,
        `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France`.longitude,
      )
      .map(_.nomCommune) shouldBe Some("LOURTIES-MONBRUN")

    Geofla
      .strTreefindBy(
        london.latitude,
        london.longitude,
      )
      .map(_.nomCommune) shouldBe None
  }
}


core/src/test/scala/com/colisweb/geoflram/GeoflaSpec.scala
package com.colisweb.geoflram

import org.scalatest.{FlatSpec, Matchers}

class GeoflaSpec extends FlatSpec with Matchers {

  "true" should "be true" in {
    true shouldBe true
  }

  "Geofla#findBy" should "find the correct commune" in {
    final case class GpsPoint(latitude: Double, longitude: Double)

    val `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France` =
      GpsPoint(latitude = 49.4296517, longitude = 1.2554941)

    val `Rue des Huttes, tertiary, Coingt, France` = GpsPoint(latitude = 49.7847695, longitude = 4.0952504)

    val `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France` =
      GpsPoint(latitude = 43.45603285, longitude = 0.5390568833333329)

    val london = GpsPoint(latitude = 51.5073219, longitude = -0.1276474)

    Geoflram
      .findBy(
        `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France`.latitude,
        `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France`.longitude,
      )
      .map(_.nomCommune) shouldBe Some("BOIS-D'ENNEBOURG")

    Geoflram
      .findBy(
        `Rue des Huttes, tertiary, Coingt, France`.latitude,
        `Rue des Huttes, tertiary, Coingt, France`.longitude,
      )
      .map(_.nomCommune) shouldBe Some("COINGT")

    Geoflram
      .findBy(
        `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France`.latitude,
        `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France`.longitude,
      )
      .map(_.nomCommune) shouldBe Some("LOURTIES-MONBRUN")

    Geoflram
      .findBy(
        london.latitude,
        london.longitude,
      )
      .map(_.nomCommune) shouldBe None
  }

}


jruby/src/test/scala/com/colisweb/geoflram/jruby/GeoflaJRubySpec.scala
package com.colisweb.geoflram.jruby

import com.colisweb.geoflram.jruby
import org.scalatest.{FlatSpec, Matchers}

class GeoflaJRubySpec extends FlatSpec with Matchers {

  "true" should "be true" in {
    true shouldBe true
  }

  "GeoflaJRuby#findBy" should "find the correct commune" in {
    final case class GpsPoint(latitude: Double, longitude: Double)

    val `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France` =
      GpsPoint(latitude = 49.4296517, longitude = 1.2554941)

    val `Rue des Huttes, tertiary, Coingt, France` = GpsPoint(latitude = 49.7847695, longitude = 4.0952504)

    val `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France` =
      GpsPoint(latitude = 43.45603285, longitude = 0.5390568833333329)

    val london = GpsPoint(latitude = 51.5073219, longitude = -0.1276474)

    jruby.GeoflramJRuby
      .findBy(
        `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France`.latitude,
        `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France`.longitude,
      )
      .nomCommune shouldBe "BOIS-D'ENNEBOURG"

    jruby.GeoflramJRuby
      .findBy(
        `Rue des Huttes, tertiary, Coingt, France`.latitude,
        `Rue des Huttes, tertiary, Coingt, France`.longitude,
      )
      .nomCommune shouldBe "COINGT"

    jruby.GeoflramJRuby
      .findBy(
        `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France`.latitude,
        `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France`.longitude,
      )
      .nomCommune shouldBe "LOURTIES-MONBRUN"

    jruby.GeoflramJRuby
      .findBy(
        london.latitude,
        london.longitude,
      ) shouldBe null
  }

}


core/src/main/scala/com/guizmaii/geofla/Geofla.scala
package com.guizmaii.geofla

import org.locationtech.jts.geom._
import org.locationtech.jts.index.strtree.STRtree
import org.locationtech.jts.io.WKTReader

import scala.io.{Codec, Source}

object Geofla {

  import scala.collection.JavaConverters._

  final case class Commune(
      geometry: Geometry,
      idGeofla: String,
      codeCommune: String,
      inseeCommune: String,
      nomCommune: String,
      statut: String,
      xChefLieu: String,
      yChefLieu: String,
      xCentroid: String,
      yCentroid: String,
      zMoyen: String,
      superficie: String,
      population: String,
      codeArr: String, // arr ?
      codeDepartement: String,
      nomDepartement: String,
      codeRegion: String,
      nomRegion: String
  )

  private[this] final val lines           = Source.fromResource("COMMUNE.csv")(Codec.UTF8).getLines().drop(1)
  private[this] final val geometryFactory = new GeometryFactory(new PrecisionModel(), 4326)
  private[this] final val reader          = new WKTReader(geometryFactory)
  private[this] final val strTree         = new STRtree()

  lines
    .map(_.split("\"").drop(1))
    .map {
      case Array(wkt: String, rest: String) => wkt :: rest.split(',').drop(1).toList
    }
    .foreach { list =>
      val commune = Commune(
        geometry = reader.read(list(0)),
        idGeofla = list(1),
        codeCommune = list(2),
        inseeCommune = list(3),
        nomCommune = list(4),
        statut = list(5),
        xChefLieu = list(6),
        yChefLieu = list(7),
        xCentroid = list(8),
        yCentroid = list(9),
        zMoyen = list(10),
        superficie = list(11),
        population = list(12),
        codeArr = list(13),
        codeDepartement = list(14),
        nomDepartement = list(15),
        codeRegion = list(16),
        nomRegion = list(17)
      )

      strTree.insert(commune.geometry.getEnvelopeInternal, commune)
    }

  strTree.build()

  def findBy(latitude: Double, longitude: Double): Option[Commune] = {
    val point: Geometry = geometryFactory.createPoint(new Coordinate(longitude, latitude))

    strTree
      .query(point.getEnvelopeInternal)
      .asInstanceOf[java.util.List[Commune]]
      .asScala
      .find(_.geometry.contains(point))
  }

}


core/src/main/scala/com/colisweb/geoflram/Geoflram.scala
package com.colisweb.geoflram

import org.locationtech.jts.geom._
import org.locationtech.jts.index.strtree.STRtree
import org.locationtech.jts.io.WKTReader

import scala.io.{Codec, Source}

object Geoflram {

  import scala.collection.JavaConverters._

  final case class Commune(
      geometry: Geometry,
      idGeofla: String,
      codeCommune: String,
      inseeCommune: String,
      nomCommune: String,
      statut: String,
      xChefLieu: String,
      yChefLieu: String,
      xCentroid: String,
      yCentroid: String,
      zMoyen: String,
      superficie: String,
      population: String,
      codeArr: String, // arr ?
      codeDepartement: String,
      nomDepartement: String,
      codeRegion: String,
      nomRegion: String
  )

  private[this] final val lines           = Source.fromResource("COMMUNE.csv")(Codec.UTF8).getLines().drop(1)
  private[this] final val geometryFactory = new GeometryFactory(new PrecisionModel(), 4326)
  private[this] final val reader          = new WKTReader(geometryFactory)
  private[this] final val strTree         = new STRtree()

  lines
    .map(_.split("\"").drop(1))
    .map {
      case Array(wkt: String, rest: String) => wkt :: rest.split(',').drop(1).toList
    }
    .foreach { list =>
      val commune = Commune(
        geometry = reader.read(list(0)),
        idGeofla = list(1),
        codeCommune = list(2),
        inseeCommune = list(3),
        nomCommune = list(4),
        statut = list(5),
        xChefLieu = list(6),
        yChefLieu = list(7),
        xCentroid = list(8),
        yCentroid = list(9),
        zMoyen = list(10),
        superficie = list(11),
        population = list(12),
        codeArr = list(13),
        codeDepartement = list(14),
        nomDepartement = list(15),
        codeRegion = list(16),
        nomRegion = list(17)
      )

      strTree.insert(commune.geometry.getEnvelopeInternal, commune)
    }

  strTree.build()

  def findBy(latitude: Double, longitude: Double): Option[Commune] = {
    val point: Geometry = geometryFactory.createPoint(new Coordinate(longitude, latitude))

    strTree
      .query(point.getEnvelopeInternal)
      .asInstanceOf[java.util.List[Commune]]
      .asScala
      .find(_.geometry.contains(point))
  }

}


benchmarks/src/main/scala/com/colisweb/geoflram/benchmark/FindByLatitudeLongitudeBenchmark.scala
package com.colisweb.geoflram.benchmark

import java.util.concurrent.TimeUnit

import com.colisweb.geoflram.Geoflram
import com.colisweb.geoflram.Geoflram.Commune
import org.openjdk.jmh.annotations._

object FindByLatitudeLongitudeBenchmark {

  final case class GpsPoint(latitude: Double, longitude: Double)

  // GpsPoints found thanks to https://www.maps.ie/coordinates.html

  val `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France` = GpsPoint(latitude = 49.4296517, longitude = 1.2554941)

  val `Rue des Huttes, tertiary, Coingt, France` = GpsPoint(latitude = 49.7847695, longitude = 4.0952504)

  val `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France` =
    GpsPoint(latitude = 43.45603285, longitude = 0.5390568833333329)

  val london = GpsPoint(latitude = 51.5073219, longitude = -0.1276474)

}

@Warmup(iterations = 10, time = 2000, timeUnit = TimeUnit.MILLISECONDS)
@Measurement(iterations = 10, time = 2000, timeUnit = TimeUnit.MILLISECONDS)
@Fork(1)
@BenchmarkMode(Array(Mode.AverageTime))
@OutputTimeUnit(TimeUnit.NANOSECONDS)
class FindByLatitudeLongitudeBenchmark {

  import FindByLatitudeLongitudeBenchmark._

  @Benchmark
  def firstCommuneOfTheList: Option[Commune] =
    Geoflram.findBy(
      `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France`.latitude,
      `Cimetière de Lourties-Monbrun, cemetery, Lourties-Monbrun, France`.longitude
    )

  @Benchmark
  def middleOfTheList: Option[Commune] =
    Geoflram.findBy(
      `Rue des Huttes, tertiary, Coingt, France`.latitude,
      `Rue des Huttes, tertiary, Coingt, France`.longitude
    )

  @Benchmark
  def lastCommuneOfTheList: Option[Commune] =
    Geoflram.findBy(
      `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France`.latitude,
      `Rue de la Fondance, tertiary, Bois-d'Ennebourg, France`.longitude
    )

  @Benchmark
  def notInTheList: Option[Commune] = Geoflram.findBy(london.latitude, london.longitude)

}
